synapse/types/__init__.py

```diff
@@ -60,6 +61,8 @@
 from synapse.util.stringutils import parse_and_validate_server_name
 
 if TYPE_CHECKING:
+    from typing_extensions import Self
```

Comment:
This requires `typing_extensions >= 4.0`. Not sure how we feel about bumping that? Or just leaving it behind `TYPE_CHECKING`?

Reply:
I'm happy as long as the packagers are. https://pkgs.org/search/?q=typing-extensions and https://repology.org/project/python:typing-extensions/versions. Debian buster and Ubuntu focal + jammy have 3.x; I can't remember if we've dropped support for those.

Reply:
Oh ISWYM. It'd be nice to be able to just import it, but I don't mind the conditional import and referencing it in "quotes" if that helps the packagers too. (ISTR we have to quote "defer.Deferred[blah]" anyway...

Reply:
I think putting it behind `TYPE_CHECKING` is fine.